flows/engine/session.go

```diff
@@ -51,117 +60,302 @@ func (s *session) GetRun(uuid flows.RunUUID) (flows.FlowRun, error) {
+				if currentRun.Status() == flows.RunStatusErrored {
+					s.status = flows.SessionStatusErrored
+				} else {
+					s.status = flows.SessionStatusCompleted
+				}
+
+				// return to caller
+				return nil
+			}
+		}
+
+		// if we now have a destination, go there
+		if destination != noDestination {
+			if !visited[destination] {
+				node := currentRun.Flow().GetNode(destination)
+				if node == nil {
+					return fmt.Errorf("unable to find destination node %s in flow %s", destination, currentRun.Flow().UUID())
+				}
+
+				step, destination, err = s.visitNode(currentRun, node, callerEvents)
+				if err != nil {
+					return err
+				}
+
+				// if we hit a wait, also return to the caller
+				if s.status == flows.SessionStatusWaiting {
+					return nil
+				}
+
+				// mark this node as visited to prevent loops
+				visited[node.UUID()] = true
+
+				// only pass our caller events to the first node as it is responsible for handling them
+				callerEvents = nil
+			} else {
+				// this is a loop, we log it and stop execution
+				currentRun.AddFatalError(step, nil, fmt.Errorf("flow loop detected, stopping execution before entering '%s'", destination))
+				destination = noDestination
+			}
+		}
+	}
 }
 
-func (s *session) resumeRun(run flows.FlowRun, callerEvents []flows.Event) error {
-	// no steps to resume from, nothing to do, return
-	if len(run.Path()) == 0 {
-		return nil
+// visits the given node, creating a step in our current run path
+func (s *session) visitNode(run flows.FlowRun, node flows.Node, callerEvents []flows.Event) (flows.Step, flows.NodeUUID, error) {
+	step := run.CreateStep(node)
+
+	// apply any caller events
+	for _, event := range callerEvents {
+		run.ApplyEvent(step, nil, event)
 	}
 
-	// grab the last step
-	step := run.Path()[len(run.Path())-1]
+	// execute our node's actions
+	if node.Actions() != nil {
+		for _, action := range node.Actions() {
+			if err := action.Execute(run, step); err != nil {
+				return nil, noDestination, err
+			}
 
-	// and the last node
-	node := run.Flow().GetNode(step.NodeUUID())
-	if node == nil {
-		err := fmt.Errorf("cannot resume at node '%s' that no longer exists", step.NodeUUID())
-		run.AddError(step, err)
-		run.Exit(flows.StatusErrored)
-		return nil
+			if run.Status() == flows.RunStatusErrored {
+				return step, noDestination, nil
+			}
+		}
 	}
 
-	// the first event resumes the wait
-	destination, step, err := resumeNode(run, node, step, callerEvents)
-	if err != nil {
-		return err
+	// a start flow action may have triggered a subflow in which case we're done on this node for now
+	// and it will be resumed when the subflow finishes
+	if s.flowTrigger != nil {
+		return step, noDestination, nil
 	}
 
-	err = continueRunUntilWait(run, destination, step, nil)
-	if err != nil {
-		return err
+	// if our node has a wait before its router, we hand back to the caller
+	wait := node.Wait()
```

Comment:
Feels to me like waits should have the opportunity to exit immediately. You could imagine a wait being conditional in that in some instances it does return back to caller and others it moves forward immediately based on some state of the run for example.